File: go/gildedrose/gildedrose.go
package gildedrose

import (
	"strings"
)

type Item struct {
	Name            string
	SellIn, Quality int
}

func UpdateQuality(items []*Item) {
	for _, item := range items {
		var shiftAmt int

		//Update shift amount based on Name
		switch item.Name {
		case "Sulfuras, Hand of Ragnaros":
			continue
		case "Aged Brie":
			shiftAmt = 1

		case "Backstage passes to a TAFKAL80ETC concert":
			{
				// Backstage passes gain quality as the concert date approaches and fall to 0 after the concert
				if item.SellIn > 0 && item.SellIn < 11 {
					if item.SellIn < 6 {
						shiftAmt = 3
					} else {
						shiftAmt = 2
					}
				} else {
					item.Quality = 0
				}
			}

		default:
			{
				if item.Quality > 0 {
					if strings.Contains(item.Name, "Conjured") {
						//Conjured items degrade in quality twice as fast
						shiftAmt = -2
					} else {
						shiftAmt = -1
					}
				}
			}
		}

		if item.SellIn > 0 {
			item.SellIn--
		}
		if item.SellIn == 0 && item.Name == "Backstage passes to a TAFKAL80ETC concert" {
			item.Quality = 0
			continue
		}

		if item.SellIn < 0 {
			if item.Name != "Aged Brie" && item.Quality > 0 {
				shiftAmt--
			}
		}

		item.Quality += shiftAmt

		if item.Quality > 50 {
			item.Quality = 50
		}
	}
}

File: Java/src/main/java/com/gildedrose/GildedRose.java
package com.gildedrose;

class GildedRose {
    Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void updateQuality() {
        for (int i = 0; i < items.length; i++) {
            Item item = items[i];
            String itemName = item.name;

            if (itemName.equals("Sulfuras, Hand of Ragnaros")) {
                continue;
            }

            if (isAgedProduct(itemName) && item.quality < 50) {
                addQuality(item);
            } else if (item.quality > 0) {
                // Conjured items degrade in quality twice as fast
                if (itemName.contains("Conjured")) {
                    item.quality--;
                }
                item.quality--;
            }

            item.sellIn--;

            if (item.sellIn < 0 && itemName.equals("Backstage passes to a TAFKAL80ETC concert")) {
                item.quality = 0;
            }
        }
    }

    public boolean isAgedProduct(String itemName) {
        return (itemName.equals("Aged Brie")
                || itemName.equals("Backstage passes to a TAFKAL80ETC concert"));
    }

    public void addQuality(Item item) {
        if (item.name.equals("Aged Brie")) {

            item.quality++;

        } else if (item.name.equals("Backstage passes to a TAFKAL80ETC concert")) {
            item.quality++;
            if (item.sellIn < 11 && item.quality < 50)
                item.quality++;

            if (item.sellIn < 6 && item.quality < 50)
                item.quality++;
        }
    }

}
